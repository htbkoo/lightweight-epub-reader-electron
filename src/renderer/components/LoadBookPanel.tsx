import * as React from 'react';
import clsx from 'clsx';
import {Book, createSimplifiedToTraditionalConverter, readEpub} from "epub-chinese-converter";
import Button from '@material-ui/core/Button';
import {createStyles, makeStyles, Theme} from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';

import {BookState} from "../reducers/bookReducer";
import {ButtonMouseEvent} from './types';
import {AppState} from "../reducers/appReducer";

// const converter = createSimplifiedToTraditionalConverter();

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            flexGrow: 1,
        },
        menuButton: {
            marginRight: theme.spacing(2),
        },
        title: {
            flexGrow: 1,
        },
        input: {
            display: 'none',
        },
        container: {
            display: "flex",
            alignItems: "center",
        },
        inline: {
            display: "inline"
        }
    }),
);

export interface Props {
    app: AppState;
    book: BookState;
    setBookContent: (book: Book.BookWithMeta) => any;
    setBookError: (error: unknown) => any;
    setFileName: (fileName: string) => any;
    notifyLoadingBook: () => any;
    setDrawerOpen: (open: boolean) => any;
}

const LoadBookPanel = ({app, book, notifyLoadingBook, setFileName, setBookContent, setBookError, setDrawerOpen}: Props) => {
    const classes = useStyles();
    return (
        <div className={clsx(classes.root, "form-group")}>
            {/*TODO: investigate if `form` is necessary and clean up if not*/}
            <form className="form-horizontal">
                <div className={classes.container}>
                    <div className={classes.inline}>
                        <IconButton
                            edge="start"
                            className={classes.menuButton}
                            color="inherit"
                            aria-label="menu"
                            onClick={handleMenuButtonClick}
                        >
                            <MenuIcon/>
                        </IconButton>
                    </div>

                    <div className={clsx(classes.inline, classes.title)}>
                        <EpubFilePicker onFilePathChange={handleFilePathChange} book={book}/>
                    </div>

                    <div className={classes.inline}>
                        <Button variant="contained" onClick={handleTranslateButtonClick} size="small" color="secondary">
                            To Traditional Chinese
                        </Button>
                    </div>
                </div>
            </form>
        </div>
    );

    function handleMenuButtonClick(e: ButtonMouseEvent) {
        e.preventDefault();

        // TODO: move this to redux store / selector to keep the state properly
        const isBookLoaded = book.bookWithMeta;

        const shouldOpenBookmarkDrawer = isBookLoaded && !app.isBookmarkDrawerOpen;
        return setDrawerOpen(shouldOpenBookmarkDrawer);
    }

    function handleTranslateButtonClick(e: ButtonMouseEvent) {
        e.preventDefault();
        if (book.bookWithMeta) {
            // setBookContent(converter.convertBook(book.bookWithMeta))
        }
    }

    function handleFilePathChange(bookUrl: string) {
        notifyLoadingBook();
        setFileName(bookUrl);
        // return readEpub(bookUrl).then(setBookContent).catch(setBookError);
    }
};

function EpubFilePicker({book, onFilePathChange}: { book: BookState, onFilePathChange: (filePath: string) => void }) {
    const classes = useStyles();

    return (
        <>
            <input className={classes.input} type="file" id="file-path-input" onClick={handleFileButtonClick}/>
            <label htmlFor="file-path-input">
                <Button variant="contained" component="span" size="small" color="secondary">
                    {getLoadButtonText(book)}
                </Button>
            </label>
        </>
    );

    function handleFileButtonClick(evt: ButtonMouseEvent) {
        evt.preventDefault();
          ipcApi.openEpubFile()
          .then((filePaths) => {
              if (filePaths === undefined) { return } // Dialog was cancelled

              console.log(`opened: ${JSON.stringify(filePaths)}`);
              return onFilePathChange(filePaths[0]);
          })
    }
}

function getLoadButtonText(book: BookState) {
    if (book.isLoadingBook) {
        return `Loading`;
    } else if (book.bookWithMeta || book.fileName) {
        const array = [];

        // TODO: migrate to something like safe get
        if (book.bookWithMeta && book.bookWithMeta.metadata) {
            if ('title' in book.bookWithMeta.metadata) {
                array.push(book.bookWithMeta.metadata.title);
            }
            if ('creator' in book.bookWithMeta.metadata) {
                array.push(book.bookWithMeta.metadata.creator);
            }
        } else if (book.fileName) {
            array.push(book.fileName);
        }
        return array.join(' - ');
    } else {
        return `Choose an epub file`;
    }
}

export default LoadBookPanel;

