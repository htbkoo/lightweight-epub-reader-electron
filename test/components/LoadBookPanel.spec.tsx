import * as React from 'react';
import * as renderer from 'react-test-renderer';

import LoadBookPanel, {Props} from '../../src/renderer/components/LoadBookPanel';
import {createAppState, createBookState} from "../utils/testUtils";

describe('<LoadBookPanel/>', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(createLoadBookPanel())
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // TODO: investigate, this test is broken even after updating snapshot
    xit('renders when isLoadingBook', () => {
        const tree = renderer
            .create(createLoadBookPanel({book: createBookState({isLoadingBook: true})}))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // TODO: investigate, this test is broken even after updating snapshot
    xit('renders with only fileName available', () => {
        const tree = renderer
            .create(createLoadBookPanel({book: createBookState({fileName: "someFileName"})}))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // TODO: investigate, this test is broken even after updating snapshot
    xit('renders with only book metadata available', () => {
        const tree = renderer
            .create(createLoadBookPanel({
                book: createBookState({
                    bookWithMeta: {
                        metadata: {creator: "creator", title: "title"},
                        chapters: {}
                    }
                })
            }))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    // TODO: investigate, this test is broken even after updating snapshot
    xit('renders with both fileName and book metadata available', () => {
        const tree = renderer
            .create(createLoadBookPanel({
                book: createBookState({
                    bookWithMeta: {
                        metadata: {creator: "creator", title: "title"},
                        chapters: {}
                    }, fileName: "someFileName"
                })
            }))
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    function createLoadBookPanel(overrides: Partial<Props> = {}) {
        const props: Props = {
            setFileName: jest.fn(),
            setBookContent: jest.fn(),
            setBookError: jest.fn(),
            notifyLoadingBook: jest.fn(),
            setDrawerOpen: jest.fn(),
            book: createBookState(),
            app: createAppState(),
            ...overrides
        };

        return <LoadBookPanel {...props}/>;
    }
});
