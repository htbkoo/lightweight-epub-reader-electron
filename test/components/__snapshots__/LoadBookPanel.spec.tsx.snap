// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadBookPanel/> renders correctly 1`] = `
<div
  className="makeStyles-root-1 form-group"
>
  <form
    className="form-horizontal"
  >
    <div
      className="makeStyles-container-5"
    >
      <div
        className="makeStyles-inline-6"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="makeStyles-inline-6 makeStyles-title-3"
      >
        <input
          className="makeStyles-input-4"
          id="file-path-input"
          onClick={[Function]}
          type="file"
        />
        <label
          htmlFor="file-path-input"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              Choose an epub file
            </span>
          </span>
        </label>
      </div>
      <div
        className="makeStyles-inline-6"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            To Traditional Chinese
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`<LoadBookPanel/> renders when isLoadingBook 1`] = `
<div
  className="makeStyles-root-1 form-group"
>
  <form
    className="form-horizontal"
  >
    <div
      className="makeStyles-container-5"
    >
      <div
        className="makeStyles-inline-6"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="makeStyles-inline-6 makeStyles-title-3"
      >
        <input
          className="makeStyles-input-4"
          id="file-path-input"
          onClick={[Function]}
          type="file"
        />
        <label
          htmlFor="file-path-input"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              Loading
            </span>
          </span>
        </label>
      </div>
      <div
        className="makeStyles-inline-6"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            To Traditional Chinese
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`<LoadBookPanel/> renders with both fileName and book metadata available 1`] = `
<div
  className="makeStyles-root-1 form-group"
>
  <form
    className="form-horizontal"
  >
    <div
      className="makeStyles-container-5"
    >
      <div
        className="makeStyles-inline-6"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="makeStyles-inline-6 makeStyles-title-3"
      >
        <input
          className="makeStyles-input-4"
          id="file-path-input"
          onClick={[Function]}
          type="file"
        />
        <label
          htmlFor="file-path-input"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              title - creator
            </span>
          </span>
        </label>
      </div>
      <div
        className="makeStyles-inline-6"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            To Traditional Chinese
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`<LoadBookPanel/> renders with only book metadata available 1`] = `
<div
  className="makeStyles-root-1 form-group"
>
  <form
    className="form-horizontal"
  >
    <div
      className="makeStyles-container-5"
    >
      <div
        className="makeStyles-inline-6"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="makeStyles-inline-6 makeStyles-title-3"
      >
        <input
          className="makeStyles-input-4"
          id="file-path-input"
          onClick={[Function]}
          type="file"
        />
        <label
          htmlFor="file-path-input"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              title - creator
            </span>
          </span>
        </label>
      </div>
      <div
        className="makeStyles-inline-6"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            To Traditional Chinese
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`<LoadBookPanel/> renders with only fileName available 1`] = `
<div
  className="makeStyles-root-1 form-group"
>
  <form
    className="form-horizontal"
  >
    <div
      className="makeStyles-container-5"
    >
      <div
        className="makeStyles-inline-6"
      >
        <button
          aria-label="menu"
          className="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="makeStyles-inline-6 makeStyles-title-3"
      >
        <input
          className="makeStyles-input-4"
          id="file-path-input"
          onClick={[Function]}
          type="file"
        />
        <label
          htmlFor="file-path-input"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiButton-label"
            >
              someFileName
            </span>
          </span>
        </label>
      </div>
      <div
        className="makeStyles-inline-6"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            To Traditional Chinese
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;
